### PR TITLE
Replace evented dataclasses with pydantic evented model on viewer

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
[EDITED] This PR replaces the `evented_dataclass` with the Pydantic `EventedModel` for all our objects on the ViewerModel and the Colormap.

It is currently a breaking change because `ndim` in `Dims` is now set to become keyword only. I've made it so that `colors` in `Colormap` can still be a non-kwarg too

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by sofroniewn** [comment]
@tlambert03 I figured out why all the Qt tests are failing, it is because of my "evented decorator" approach, which looks like it is being applied to the class rather than the instance.

For example

```python
from napari.components import Dims


d1 = Dims()
d2 = Dims()

d1.events == d2.events
```

which is not what we want! Does that surprise you? Any ideas how to fix? I'll look in now too

**Comment by tlambert03** [comment]
> Does that surprise you?

nope, that's how decorators work!
```python
@decorator
class Dims:
	pass

# is the same as 

class Dims:
    pass
Dims = decorator(Dims)
# (you're just redefining the class with the output of decorator)
```

the code you borrowed there to add `events` is [normally inside of a `_post_init_` method that I add to the class](https://github.com/napari/napari/blob/cc3ed86629b9e448f6d8622a82979d21c393d943/napari/utils/events/dataclass.py#L314-L331), whereas your `evented_model` is adding `events` directly as a class attribute

**Comment by sofroniewn** [comment]
Yup makes sense! I thought the decorator approach was so slick though. Any ideas how to save it, or is it back to mixin/ something in init?

File: napari/_qt/_tests/test_qt_viewer.py
Comment (sofroniewn) [inline napari/_qt/_tests/test_qt_viewer.py]: Moving this wasn't strictly necessary, but it was a test using the whole viewer in qt_dims. confused me up while debugging. could move it back, but hopefully people are ok with move

File: napari/_qt/widgets/_tests/test_qt_dims.py
Comment (sofroniewn) [inline napari/_qt/widgets/_tests/test_qt_dims.py:18]: pydantic models are kwarg only so `Dims(ndim)` no longer works, needs to be `Dims(ndim=ndim)`

File: napari/_qt/widgets/qt_dims.py
Comment (sofroniewn) [inline napari/_qt/widgets/qt_dims.py:133]: For typing reasons we now have this always be an `int` instead of sometimes be `None`. I think it's better anyway. Could be reverted if people really wanted but I prefer this behavior

File: napari/_qt/widgets/_tests/test_qt_dims.py
Comment (tlambert03) [reply napari/_qt/widgets/_tests/test_qt_dims.py:18]: I saw that ... that's kind of a bummer, no more `Image(np.random.rand(10, 10))`?  might need to patch that...

**Comment by sofroniewn** [comment]
Ok, thanks to some great help from @tlambert03 all tests are now passing apart from our minreq one, which doesn't even get started. It fails right away with

```python-traceback
  ________________________ ERROR collecting test session _________________________
  /opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/importlib/__init__.py:127: in import_module
      return _bootstrap._gcd_import(name[level:], package, level)
  <frozen importlib._bootstrap>:1006: in _gcd_import
      ???
  <frozen importlib._bootstrap>:983: in _find_and_load
      ???
  <frozen importlib._bootstrap>:953: in _find_and_load_unlocked
      ???
  <frozen importlib._bootstrap>:219: in _call_with_frames_removed
      ???
  <frozen importlib._bootstrap>:1006: in _gcd_import
      ???
  <frozen importlib._bootstrap>:983: in _find_and_load
      ???
  <frozen importlib._bootstrap>:967: in _find_and_load_unlocked
      ???
  <frozen importlib._bootstrap>:677: in _load_unlocked
      ???
  <frozen importlib._bootstrap_external>:728: in exec_module
      ???
  <frozen importlib._bootstrap>:219: in _call_with_frames_removed
      ???
  napari/__init__.py:7: in <module>
      from .viewer import Viewer  # isort:skip
  napari/viewer.py:2: in <module>
      from .components import ViewerModel
  napari/components/__init__.py:17: in <module>
      from .camera import Camera
  napari/components/camera.py:3: in <module>
      from pydantic import validator
  pydantic/__init__.py:2: in init pydantic.__init__
      ???
  pydantic/dataclasses.py:7: in init pydantic.dataclasses
      ???
  pydantic/main.py:333: in init pydantic.main
      ???
  pydantic/main.py:326: in pydantic.main.ModelMetaclass.__new__
      ???
  E   AttributeError: attribute '__signature__' of 'type' objects is not writable
```

I'm not quite sure what's going on here. Some googling does turn up stuff though, so it should be possible to figure out.

I'd like to get #2037 merged and then clean this one up before final review. I should also add some more tests, doc strings etc. but I'm feeling very good about this

**Comment by tlambert03** [comment]
> ```
>  AttributeError: attribute '__signature__' of 'type' objects is not writable
> ```

here's your problem: https://bugreports.qt.io/browse/PYSIDE-1004
bumping our minimum PySide2 requirement to 5.13.1 (from 5.12.3) would fix it... may be another workaround

**Comment by tlambert03** [comment]
> may be another workaround

eh... I don't think so.  This would probably mean that we have to drop support for PySide2<5.13.1

**Comment by sofroniewn** [comment]
> eh... I don't think so. This would probably mean that we have to drop support for PySide2<5.13.1

Ok good to know. I'm ok with that for what I see as such a large gain here, what do you think @tlambert03?

We're at 5.15.2 right now and it's been over a year since `5.13.1` came out. `5.12.3` also isn't that much older, so I don't think it will effect too many people.

Also if you wanted to give this a first pass review that would be welcome - I'd say for now fine to focus only on things that would be "blocker to merge" status, other stuff can come later, though all feedback welcome.

**Comment by tlambert03** [comment]
did a quick `setattr` perf test on "Old" (evented_dataclass) vs "New" (pydantic).  Definitely no problems, looks to be faster if anything! 

```
---------------------------------------------- benchmark: 8 tests ---------------------------------------------------------------
Name (time in us)        Min                Max              Mean            StdDev            Median          Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------------------
test_enum[New]        3.4170 (1.38)      7.0396 (1.50)     3.5852 (1.37)     0.1503 (2.35)     3.5663 (1.37)    10000         100
test_enum[Old]        4.8218 (1.95)     20.7601 (4.43)     5.0516 (1.94)     0.4431 (6.94)     4.9867 (1.92)    10000         100
test_int[New]         2.4693 (1.0)       4.6842 (1.0)      2.6088 (1.0)      0.0990 (1.55)     2.5960 (1.0)     10000         100
test_int[Old]         4.6018 (1.86)      9.6662 (2.06)     4.8106 (1.84)     0.2145 (3.36)     4.7761 (1.84)    10000         100
test_str[New]         2.5919 (1.05)      5.1877 (1.11)     2.7247 (1.04)     0.0638 (1.0)      2.7282 (1.05)    10000         100
test_str[Old]         4.7323 (1.92)      7.0208 (1.50)     4.8801 (1.87)     0.0854 (1.34)     4.8739 (1.88)    10000         100
test_tuple[New]       2.5824 (1.05)      5.0099 (1.07)     2.7329 (1.05)     0.1466 (2.30)     2.7113 (1.04)    10000         100
test_tuple[Old]       4.7136 (1.91)      7.9404 (1.70)     4.9245 (1.89)     0.1034 (1.62)     4.9191 (1.89)    10000         100
---------------------------------------------------------------------------------------------------------------------------------
```

File: napari/utils/_pydantic/custom_types.py
Comment (tlambert03) [inline napari/utils/_pydantic/custom_types.py]:
```suggestion
class Array(np.ndarray):
    def __class_getitem__(cls, t):
        return type('Array', (Array,), {'__dtype__': t})
```

This metaclass strategy would also work with <py3.7, but since 3.7 we have [`__class_getitem__`](https://www.python.org/dev/peps/pep-0560/#specification) which looks better, and lets us avoid using metaclasses and the conflicts that they sometimes cause.

**Review by tlambert03 — COMMENTED** [review]
just starting to look, but so far I love it!

File: napari/utils/_pydantic/evented_model.py
Comment (tlambert03) [inline napari/utils/_pydantic/evented_model.py]:
```suggestion
        # whether to allow arbitrary user types for fields (they are validated
        # simply by checking if the value is an instance of the type). If
        # False, RuntimeError will be raised on model declaration
        arbitrary_types_allowed = True
        # whether to perform validation on assignment to attributes
        validate_assignment = True
        # whether to treat any underscore non-class var attrs as private
        # https://pydantic-docs.helpmanual.io/usage/models/#private-model-attributes
        underscore_attrs_are_private = True
        # whether to populate models with the value property of enums, rather
        # than the raw enum. This may be useful if you want to serialise
        # model.dict() later
        use_enum_values = True
        # whether to validate field defaults (default: False)
        validate_all = True
        # a dict used to customise the way types are encoded to JSON
        # https://pydantic-docs.helpmanual.io/usage/exporting_models/#modeljson
        json_encoders = JSON_ENCODERS
```

Comment (tlambert03) [reply napari/utils/_pydantic/evented_model.py]: lets just comment the crap out of EventedModel, so we don't have to go to pydantic docs each time

File: napari/utils/_pydantic/evented_model.py
Comment (tlambert03) [inline napari/utils/_pydantic/evented_model.py]:
```suggestion
    # pydantic BaseModel configuration.  see:
    # https://pydantic-docs.helpmanual.io/usage/model_config/
```

File: napari/utils/_pydantic/evented_model.py
Comment (tlambert03) [inline napari/utils/_pydantic/evented_model.py]: class docstring

File: napari/utils/_pydantic/evented_model.py
Comment (tlambert03) [inline napari/utils/_pydantic/evented_model.py]:
```suggestion
        # add events for each field
        self._events.source = self
        self._events.add(**dict.fromkeys(self.__fields__))
```
note, it's important that `self._events.source = self` come before adding events.  Also, the reason the everything was fine without the `__weakref__` classVar is because no events had sources anymore (which we definitely want, right?)

Comment (tlambert03) [reply napari/utils/_pydantic/evented_model.py]: but also ... I _love_ how much cleaner this handles inheritance (compared to `evented_dataclass`) with _events as a `PrivateAttr`

File: napari/utils/_pydantic/evented_model.py
Comment (tlambert03) [inline napari/utils/_pydantic/evented_model.py]:
```suggestion
    __slots__: ClassVar[Set[str]] = {"__weakref__"}

```

File: napari/utils/_pydantic/evented_model.py
Comment (tlambert03) [inline napari/utils/_pydantic/evented_model.py]:
```suggestion
from typing import ClassVar, Dict, Set
```

File: napari/utils/_pydantic/evented_model.py
Comment (tlambert03) [reply napari/utils/_pydantic/evented_model.py]: (see above... it only seemed to be ok without this because your EmitterGroup wasn't getting assigned a source)

**Review by tlambert03 — COMMENTED** [review]
adding some more comments.  but still loving it! 
Inheritance becomes much easier... specifically, it seems to avoid [this nightmare](https://stackoverflow.com/questions/51575931/class-inheritance-in-python-3-7-dataclasses)

File: napari/utils/_pydantic/evented_model.py
Comment (tlambert03) [reply napari/utils/_pydantic/evented_model.py]: (for the `__weakref__` in `__slots__`)

File: napari/components/dims.py
Comment (tlambert03) [inline napari/components/dims.py:68]: # 😍

File: napari/utils/_pydantic/custom_types.py
Comment (sofroniewn) [reply napari/utils/_pydantic/custom_types.py]: wow, that's great!

File: napari/utils/_pydantic/evented_model.py
Comment (sofroniewn) [reply napari/utils/_pydantic/evented_model.py]: Do we need a `source`? I guess might as well ... ??

Comment (tlambert03) [reply napari/utils/_pydantic/evented_model.py]: Yeah, i think we definitely dont want to lose that.  Things like nested evented list take good advantage of that attribute...

Comment (sofroniewn) [reply napari/utils/_pydantic/evented_model.py]: sounds good, thanks for the fix!

File: napari/_qt/widgets/_tests/test_qt_dims.py
Comment (tlambert03) [reply napari/_qt/widgets/_tests/test_qt_dims.py:18]: turns out this works for positional args:

```python
from pydantic import BaseModel

class TakesArgs(BaseModel):
    x: int
    y: int=4

    def __init__(self, x, **data):
        super().__init__(x=x, **data)
```

Comment (sofroniewn) [reply napari/_qt/widgets/_tests/test_qt_dims.py:18]: Ok nice that's great! I think allowing `Colormaps(color)` is nice as that is public facing. I'm fine with `Dims(ndim=2)` though as that is highly internal. Will see how others feel during review

**Comment by tlambert03** [comment]
hey @sofroniewn, how would you feel about pulling out just the `EventedModel` class into a new PR that we could potentially merge (un-hooked-up), so that it's easier test in various PRs?

**Comment by sofroniewn** [comment]
Yes, sounds great to me - could add tests just for that part too - I am away from my keyboard right now, but feel free to start a fresh PR if you want to start on it now - don’t worry about commit history

**Comment by sofroniewn** [comment]
@tlambert03 I just did this now - https://github.com/napari/napari/pull/2127 - more still to do, but we can continue discussion there

**Comment by sofroniewn** [comment]
Ok, I've updated this to use #2127, still maybe have one or two things to check over, and a test or two to fix.

Final question though is if for `Dims` and `Colormap` we should allow non-kwarg arguments, see discussion https://github.com/napari/napari/pull/2042#discussion_r550823032. It is very easy to do. Right now `Dims` can take `ndim` and `Colormap` can take `colors`. My vote is to force `ndim` to be kwarg as it is largely internal. I'm not sure about `Colormap`, I think having `colors` be non-kwarg is nice. Thoughts on this @tlambert03 @jni ?

**Comment by sofroniewn** [comment]
Test now fix, this is ready for final review - see though last comment above

> Final question though is if for Dims and Colormap we should allow non-kwarg arguments, see discussion #2042 (comment). It is very easy to do. Right now Dims can take ndim and Colormap can take colors. My vote is to force ndim to be kwarg as it is largely internal. I'm not sure about Colormap, I think having colors be non-kwarg is nice. Thoughts on this @tlambert03 @jni ?

File: napari/_qt/widgets/_tests/test_qt_dims.py
Comment (tlambert03) [reply napari/_qt/widgets/_tests/test_qt_dims.py:18]: is this PR just waiting for review?  Were you going to modify any of these `__init__` methods to allow position arguments in some cases?

Comment (sofroniewn) [reply napari/_qt/widgets/_tests/test_qt_dims.py:18]: yup, waiting for review/ any one else to weigh in with a strong opinion on which of the `__ini__` to modify or not - I'm leaning to not modifying `Dims` as it is so internal, but modifying `Coloromap` as someone else might want to use it. If you like that then I'll do it now

Comment (sofroniewn) [reply napari/_qt/widgets/_tests/test_qt_dims.py:18]: I have now done this

**Comment by sofroniewn** [comment]
Ok, I have made `Colormaps` take `colors` as a non-kwarg

File: napari/utils/colormaps/colormap.py
Comment (kevinyamauchi) [inline napari/utils/colormaps/colormap.py:26]: @sofroniewn , do you think it's worth adding `__eq__` methods so that `cmap_1 == cmap_2` works? This is useful for when an `EventedModel` has a `Colormap` as a field (i.e., for the [equality check for emitting an event when setting the field value](https://github.com/napari/napari/blob/538c1e4fee9bd272ea193511679568bd25117375/napari/utils/events/evented_model.py#L70)).~~

This might apply for the other models in this PR, but I came across this one while working on the `ColorManager` PR.

Here's a minimal example of the issue:
```python
from pydantic import validator

from napari.utils.colormaps.colormap_utils import ensure_colormap
from napari.utils.colormaps import Colormap
from napari.utils.events.evented_model import EventedModel

class MyModel(EventedModel):
    colormap: Colormap = 'viridis'

    @validator('colormap', pre=True)
    def _ensure_colormap(cls, v):
        return ensure_colormap(v)

model = MyModel(colormap='viridis')
model.colormap = 'gray'
```
gives the following warning: 

```python
.../napari/napari/utils/events/dataclass.py:223: UserWarning: Comparison method failed. Returned False. There may be need to define custom compare methods in __equality_checks__ dictionary. Exception The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
  "Comparison method failed. Returned False. "
```

Comment (sofroniewn) [reply napari/utils/colormaps/colormap.py:26]: Yes! It sounds like we need to add this because of the `Array` property in the Colormap. We already have a `__equality_checks__` dictionary, but I think we just need to wire things up properly. I can take a look soon cc @tlambert03

File: napari/_vispy/vispy_camera.py
Comment (jni) [inline napari/_vispy/vispy_camera.py]: @sofroniewn can you clarify what's happening here? I find it rather disconcerting. Is center now always 3D and we have a dummy 0.0 when we only want 2D? Is this necessary behaviour? (ie could we instead use a Union type?) If so, I think the following code might be clearer, because there's only one place where we clip and append:

```python
if self._view.camera == self._3D_camera:
    center = tuple(self.view.camera.center)
else:
    # in 2D, we arbitrarily choose 0.0 as the center in z
    center = tuple(self._view.camera.center[:2]) + (0.0,)
# switch from VisPy xyz ordering to NumPy prc ordering
center = center[::-1]
return center
```

One question I have is why this value would ever become something different than 0, if this is a different camera object? ie why are we clipping with the `[:2]` and then adding (0.0,) vs keeping that value pinned at 0?

File: napari/components/dims.py
Comment (jni) [reply napari/components/dims.py:68]: ditto!

File: napari/components/dims.py
Comment (jni) [inline napari/components/dims.py:86]: Could you add a docstring here? This function is doing a lot of work and as someone not familiar with pydantic and root_validator, I just have no idea what it's doing. Comments for each of the steps would be nice too, e.g.:

```python
# check that there is a range for each dimension.
# if there are too few ranges, prepend (0, 2, 1) for each missing dimension.
# if there are too many, throw out the leading ranges.
# this is done because ... (🤷)
```

Comment (jni) [reply napari/components/dims.py:86]: I realise looking below that this was in some sense existing code... But it's never too late to improve our docs! ;)

File: napari/components/viewer_model.py
Comment (jni) [inline napari/components/viewer_model.py:387]: Why is this change necessary?

File: napari/layers/base/base.py
Comment (jni) [inline napari/layers/base/base.py:970]: This is fun! 😂

File: napari/utils/colormaps/_tests/test_colormap.py
Comment (jni) [inline napari/utils/colormaps/_tests/test_colormap.py]: Do you want to revert this now that (as I read the conversation) you've re-added positional args to Colormap? It'd be nice to keep testing that that works.

**Review by jni — APPROVED** [review]
@sofroniewn LGTM! I've asked for very minor comments or modifications, otherwise 👍! Onwards! 🚀

File: napari/utils/colormaps/colormap.py
Comment (sofroniewn) [reply napari/utils/colormaps/colormap.py:26]: Note a PR to address this has been started in #2200

File: napari/_vispy/vispy_camera.py
Comment (sofroniewn) [reply napari/_vispy/vispy_camera.py]: Yeah, center is always length three, but we don't actually need this change here - even if we remove it when you do `camera.center = [20, 10]` say you'd get the value to be `[0, 20, 10]`. I prefer your code snippet though and will update

File: napari/components/dims.py
Comment (sofroniewn) [reply napari/components/dims.py:86]: comments and doc string added

File: napari/components/viewer_model.py
Comment (sofroniewn) [reply napari/components/viewer_model.py:387]: it's something that we should have been doing anyway - not critical for this PR to work, but otherwise when adding layers the cursor position could get the wrong dimensionality and be off.

File: napari/utils/colormaps/_tests/test_colormap.py
Comment (sofroniewn) [reply napari/utils/colormaps/_tests/test_colormap.py]: done, i reverted most tests, but missed this one

**Comment by sofroniewn** [comment]
Ok, I have addressed all reviewer comments and would like to merge after tests pass. @tlambert03 do you have any final comments on this before it goes in. I think we can continue to work on things like #2200 after merge. I don't see that as blocking.

**Pull request merged**
